spec/fixtures/generate_external_certs.sh

```diff
@@ -0,0 +1,322 @@
+  ROOT_DIR=/tmp/ca_gen
+  CSR_DIR=$ROOT_DIR/csr
+  DIR_CERTS=$ROOT_DIR/certs
+  DIR_CONF=$ROOT_DIR/conf
+  DIR_KEYS=$ROOT_DIR/keys
+  DIR_CADATA=$ROOT_DIR/cadata
+  
+  mkdir -p $CSR_DIR
+  mkdir -p $DIR_CERTS
+  mkdir -p $DIR_CONF
+  mkdir -p $DIR_KEYS
+  mkdir -p $DIR_CADATA
+}
+
+function cleanup() {
+  rm -rf $ROOT_DIR
+}
+
+function init_ca() { # $1 ca_name
+  echo "Initiating CA  \"$1\""
+  generate_conf $1
+  # Create / clear certs directory for CA
+  mkdir -p $DIR_CADATA/$1/db.certs
+  rm $DIR_CADATA/$1/db.certs/* 2>/dev/null
+  
+  # Delete both db.index and db.index.attr if they exist
+  rm $DIR_CADATA/$1/db.inde* 2>/dev/null
+  
+  # Create blank db.index to start off CA empty
+  touch $DIR_CADATA/$1/db.index
+
+  # Create db.index.attr with default contents
+  echo unique_subject = yes > $DIR_CADATA/$1/db.index.attr
+
+  # Serial can start at 1 because serials are unique per CA
+  echo 01 > $DIR_CADATA/$1/db.serial
+  
+  # Create the keyfile for the CA
+  openssl genrsa -out $DIR_KEYS/ca_$1.key 2>/dev/null
+}
+
+function init_root_ca() { # $1 ca_name, $2 subject
+  init_ca $1
+  echo "Generating CRT \"$1\""
+  openssl req -new -x509 -days 100 -extensions v3_ca -key $DIR_KEYS/ca_$1.key -out $DIR_CERTS/ca_$1.crt -subj "$2"
+}
+
+function init_sub_ca() { # $1 ca_name, $2 parent_ca, $3 subject
+  init_ca $1
+  echo "Generating CRT \"$1\""
+  openssl req -sha256 -new -key $DIR_KEYS/ca_$1.key -out $CSR_DIR/ca_$1.csr -subj "$3"
+  openssl ca -batch \
+    -config $DIR_CONF/ca_$2.conf \
+    -extensions v3_ca \
+    -out $DIR_CERTS/ca_$1.crt \
+    -infiles $CSR_DIR/ca_$1.csr 2>/dev/null
+}
+
+function append_altnames() { # $1 altnames, $2 filename
+  # Split up dash delimited string into array
+  IFS='-' read -r -a altnames <<< "$1"
+  for index in "${!altnames[@]}"
+  do
+      echo "DNS.$(($index+1)) = ${altnames[index]}" >> $2
+  done
+}
+
+function append_ip_altnames() { # $1 altnames, $2 filename
+  # Split up dash delimited string into array
+  IFS='-' read -r -a altnames <<< "$1"
+  for index in "${!altnames[@]}"
+  do
+      echo "IP.$(($index+1)) = ${altnames[index]}" >> $2
+  done
+}
+
+function site_request_conf() { # $1 file, $2 bits
+  echo "[req]" > $1
+  echo default_bits = $2 >> $1
+  echo distinguished_name = req_distinguished_name >> $1
+  echo req_extensions = req_ext >> $1
+  
+  # Various prompts for the portions of the subject would go here, as well as defaults.
+  # They can, though, be left blank when the subject is provided on the command line.
+  cat >>$1 <<EOL
+[req_distinguished_name]
+EOL
+  
+  echo -e "[no_ext]\nsubjectKeyIdentifier = hash\n" >> $1
+  echo -e "[req_ext]\nsubjectAltName = @alt_names\n" >> $1
+  echo -e "[alt_names]" >> $1
+}
+
+function gen_start_end() { # $1 good/expired
+  # Right now we are just using hardcoded good/bad for the date range validity.
+  # For most purposes it would be better to expose the clean text specification
+  # of both start and end as parameters.
+  if [ $1 = "good" ]
```

Comment:
Generally, shell comparisons using bash should use `[[ ]]` instead of `[`, because the single bracket is implemented in bash as an alias for `test` while the `[[` is implemented as a first-class construct (or whatever it's called).  The single brackets are only good when maintaining compatibility with baseline POSIX sh, which this script already doesn't do. :D

Reply:
Changed to using double brackets on all the if conditions. Also changed this specific if to use double equals as well.